authorization/src/main/kotlin/org/modelix/authorization/PermissionManagementPage.kt

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2024.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.modelix.authorization
+
+import io.ktor.server.application.ApplicationCall
+import io.ktor.server.application.application
+import io.ktor.server.application.call
+import io.ktor.server.application.plugin
+import io.ktor.server.html.respondHtml
+import io.ktor.server.request.receiveParameters
+import io.ktor.server.response.respond
+import io.ktor.server.routing.Route
+import io.ktor.server.routing.get
+import io.ktor.server.routing.post
+import io.ktor.server.routing.route
+import kotlinx.html.HTML
+import kotlinx.html.body
+import kotlinx.html.br
+import kotlinx.html.div
+import kotlinx.html.h1
+import kotlinx.html.head
+import kotlinx.html.hiddenInput
+import kotlinx.html.postForm
+import kotlinx.html.style
+import kotlinx.html.submitInput
+import kotlinx.html.table
+import kotlinx.html.td
+import kotlinx.html.textInput
+import kotlinx.html.th
+import kotlinx.html.tr
+import org.modelix.authorization.permissions.PermissionParts
+import org.modelix.authorization.permissions.PermissionSchemaBase
+
+fun Route.installPermissionManagementHandlers() {
+    route("permissions") {
+        get("manage") {
+            call.respondHtml {
+                buildPermissionManagementPage(call, application.plugin(ModelixAuthorization))
+            }
+        }
+        post("grant") {
+            val formParameters = call.receiveParameters()
+            val userId = formParameters["userId"]
+            val roleId = formParameters["roleId"]
+            require(userId != null || roleId != null) { "userId or roleId required" }
+            val permissionId = requireNotNull(formParameters["permissionId"]) { "permissionId not specified" }
+
+            // a user can grant his own permission to other users
+            checkPermission(PermissionParts.fromString(permissionId))
+
+            if (userId != null) {
+                application.plugin(ModelixAuthorization).config.accessControlPersistence.update {
+                    it.withGrantToUser(userId, permissionId)
+                }
+            }
+            if (roleId != null) {
```

Comment:
Is it a bug or a feature, that permission can be mapped to a user and a role if `userId` and `roleId` id are set?
At least the response text does not reflect the case.

Reply:
Same for `permissions/remove-grant`.

Reply:
It just isn't relevant. The UI only provides one of the parameters. If the worst thing that can happen, if someone writes his own request, is an inaccurate response text, then that's good enough for me. I will not make the code more complicated.

Reply:
ok